### DISCUSSION TITLE
`hoop`的调用顺序是不是决定了中间件的执行顺序

### DISCUSSION THREAD

**Discussion by xmh0511** [salvo-rs/salvo#804] [General]
```rust
Router::with_path("hello").hoop(A).hoop(B).get(hello);
```
在执行`hello`之前，中间件被执行的顺序是不是先`A`后`B`呢？这个行为没有在文档里面标出来

**Comment by chrislearn** [comment]
是的

**Reply by TnZzZHlp** [reply]
请教一下大佬如果是这样写呢 
```rust
Router::with_path("test").hoop(time).get(test1).hoop(log);
```
我在 time函数中往depot存放了一个time::Instant
然后在test1函数中sleep了3秒
然后在log函数中把depot存放的Instant取出并且求经过时间

但是执行的顺序却是 time -> log -> test1

如果我想获取整个处理流程所耗费的时间应该怎么实现呢？

求大佬解答

具体代码 
https://github.com/TnZzZHlp/test/blob/main/src/main.rs 59-90行

**Reply by chrislearn** [reply]
你可以看一下这个 log 中间件如何实现的

https://github.com/salvo-rs/salvo/blob/main/crates/extra/src/logging.rs

所有过滤器都通过了，开始依照你添加中间件的顺序执行中间件，最后执行 router 对应的 handler.

如果某个中间件想后续的中间件执行完再执行，调 ctrl.call_next

**Reply by TnZzZHlp** [reply]
感谢解答❤️
